nixos/modules/services/matrix/synapse/workers.nix

```diff
@@ -0,0 +1,198 @@
+        };
+        default = { };
+        description = "A freeform option of the workers settings, see upstream synapse documentation for options";
+      };
+    });
+
+    workerListenerType = types.submodule {
+      options = {
+        type = mkOption {
+          type = types.enum [ "http" "metrics" ];
+          description = "The type of the listener";
+          default = "http";
+        };
+        port = mkOption {
+          type = types.port;
+          description = "The TCP port to bind to";
+        };
+        bind_addresses = mkOption {
+          type = with types; listOf str;
+          description = "A list of local addresses to listen on";
+          default = [ "127.0.0.1" ];
+        };
+        tls = mkOption {
+          type = types.bool;
+          description = ''
+            Whether to enable TLS for this listener.
+            Will use the TLS key/cert specified in tls_private_key_path / tls_certificate_path.
+          '';
+          default = false;
+        };
+        x_forwarded = mkOption {
+          type = types.bool;
+          description = ''
+            Whether to use the X-Forwarded-For HTTP header as the client IP.
+            This option is only valid for an 'http' listener.
+            It is useful when Synapse is running behind a reverse-proxy.
+          '';
+          default = true;
+        };
+        resources = mkOption {
+          type = types.listOf (types.submodule {
+            options = {
+              names = mkOption {
+                type = types.listOf (types.enum [
+                  "client"
+                  "consent"
+                  "federation"
+                  "keys"
+                  "media"
+                  "metrics"
+                  "openid"
+                  "replication"
+                  "static"
+                  "webclient" # TODO: wtf?
+                ]);
+                description = "A list of resources to host on this port";
+                default = [ ];
+                example = lib.literalExpression ''[ "client" ]'';
+              };
+              compress = lib.mkEnableOption "HTTP compression for this resource";
+            };
+          });
+          default = [ { } ];
+          description = "The resources to make available on this listener";
+        };
+      };
+    };
+
+  in {
+    mainReplicationHost = mkOption {
+      type = types.str;
+      default = let
+        host = lib.head mainReplicationListener.bind_addresses;
+      in if builtins.elem host [ "0.0.0.0" "::"] then "127.0.0.1" else host;
+      defaultText = "Autoconfigured from main synapse config";
+      description = "The replication listeners IP on the main synapse process";
+    };
+
+    mainReplicationPort = mkOption {
+      type = types.port;
+      default = mainReplicationListener.port;
+      defaultText = "Autoconfigured from main synapse config";
+      description = "The replication listeners port on the main synapse process";
+    };
```

Comment:
Those options feel weird and confusing

Reply:
I agree, they're probably just an aspect of when this tried to preserve the possibilty of running it on multiple hosts. It's not really necessary to expose these as options anymore